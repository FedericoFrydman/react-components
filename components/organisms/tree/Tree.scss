@import '../../assets/styles/colors.scss';


.lsux-tree {
  border-left: 3px solid $N300;
}

.lsux-tree-row {
  align-items: center;
  display: flex;
  justify-content:space-between;
}

.lsux-tree-row-label{
  align-items:center;
  display:flex;
}

.lsux-tree-row-allow-select:hover {
  cursor: pointer;
}

.lsux-tree-row-selected {
  background-color: $N300;
}

.lsux-tree-row-highlighted {
  color: $R200;
}