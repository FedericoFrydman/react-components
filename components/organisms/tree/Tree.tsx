import { forwardRef, HTMLProps, Ref } from 'react';
import cn from 'classnames';

import { Button } from '../../molecules/button/Button';

import './Tree.scss';

export interface TreeItem {
  name: string;
  id?: number;
  selected?: boolean;
  expanded?: boolean;
  highlighted?: boolean;
  children?: TreeItem[];
}

export interface TreeProps extends Omit<HTMLProps<HTMLDivElement>, 'css' | 'onClick'> {
  /**
   * Root of the tree
   */
  root: TreeItem;
  /**
   * Allows selection
   */
  allowSelect?: boolean;
  /**
   * Selected callback
   */
  onClick?: (index: number) => void;
  /**
   * Open item callback
   */
  onExpand?: (index: number) => void;
  /*
   * Collapse item callback
   */
  onCollapse?: (index: number) => void;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const TreeComponent = (
  {
    root,
    classNames = [],
    allowSelect = true,
    onClick = null,
    onExpand = null,
    onCollapse = null,
    ...props
  }: TreeProps,
  ref: TreeProps['ref'],
) => {
  const openClick = function (e: React.MouseEvent, id: number) {
    if (onExpand) {
      onExpand(id);
      e.stopPropagation();
    }
  };

  const closeClick = function (e: React.MouseEvent, id: number) {
    if (onCollapse) {
      onCollapse(id);
      e.stopPropagation();
    }
  };

  const renderTreeItem = function (item: TreeItem) {
    const classTreeRow = 'lsux-tree-row';
    const cnTreeRow = cn(classTreeRow, {
      [`${classTreeRow}-allow-select`]: allowSelect,
      [`${classTreeRow}-selected`]: item.selected,
    });

    return (
      <>
        <div
          className={cnTreeRow}
          onClick={() => {
            if (allowSelect && onClick) {
              onClick(item.id);
            }
          }}
        >
          <div className={`${classTreeRow}-label`}>
            {item.children && item.children.length > 0 && item.expanded && (
              <Button
                variant="tertiary"
                iconLeft="nav_chevron_single_down"
                onClick={(e: React.MouseEvent) => {
                  closeClick(e, item.id);
                }}
              />
            )}
            {item.children && item.children.length > 0 && !item.expanded && (
              <Button
                variant="tertiary"
                iconLeft="nav_chevron_single_right"
                onClick={(e: React.MouseEvent) => {
                  openClick(e, item.id);
                }}
              />
            )}
            {(!item.children || item.children.length === 0) && <div style={{ height: '34px', width: '34px' }}></div>}
            <span className={cn('pl-2', { [`${classTreeRow}-highlighted`]: item.highlighted })}>{item.name}</span>
          </div>
        </div>
        <div className="pt-2" style={{ paddingLeft: '17px' }}>
          {item.children && item.expanded && (
            <div className={cn('lsux-tree', 'pl-3')} key={item.id}>
              {item.children.map(renderTreeItem)}
            </div>
          )}
        </div>
      </>
    );
  };

  return (
    <div {...props} ref={ref} className={cn(classNames)}>
      {renderTreeItem(root)}
    </div>
  );
};

/**
 * Tree Component
 */
export const Tree = forwardRef<HTMLDivElement, TreeProps>(TreeComponent);
