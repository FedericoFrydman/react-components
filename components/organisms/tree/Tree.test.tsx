import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Tree, TreeProps } from './Tree';

describe('Tree Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    testId = 'test-tree';
    wrapper = shallow(
      <Tree
        allowSelect={true}
        root={{
          children: [
            { id: 1, name: 'Nathan' },
            {
              children: [
                { id: 3, name: 'One' },
                { id: 4, name: 'Two' },
              ],
              id: 2,
              name: 'Jackie',
            },
            { id: 4, name: 'Leslie' },
          ],
          id: 0,
          name: 'Arnold Blinn',
        }}
      />,
    );
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the Tree the provided classNames', () => {
          const tree = renderTree({ classNames });

          expect(tree).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the snackbar', () => {
      expect(renderTree({ ref })).toEqual(ref.current);
    });
  });

  const renderTree = (props: Omit<TreeProps, 'root'> = {}): HTMLElement => {
    render(
      <Tree
        {...props}
        root={{
          children: [
            { id: 1, name: 'Nathan' },
            {
              children: [
                { id: 3, name: 'One' },
                { id: 4, name: 'Two' },
              ],
              id: 2,
              name: 'Jackie',
            },
            { id: 4, name: 'Leslie' },
          ],
          id: 0,
          name: 'Arnold Blinn',
        }}
        data-testid={testId}
      />,
    );

    return screen.getByTestId(testId);
  };
});
