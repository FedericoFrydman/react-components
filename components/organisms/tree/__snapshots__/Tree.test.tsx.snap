// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Component should render the component 1`] = `
<div
  className=""
>
  <div
    className="lsux-tree-row lsux-tree-row-allow-select"
    onClick={[Function]}
  >
    <div
      className="lsux-tree-row-label"
    >
      <ForwardRef(ButtonComponent)
        iconLeft="nav_chevron_single_right"
        onClick={[Function]}
        variant="tertiary"
      />
      <span
        className="pl-2"
      >
        Arnold Blinn
      </span>
    </div>
  </div>
  <div
    className="pt-2"
    style={
      Object {
        "paddingLeft": "17px",
      }
    }
  />
</div>
`;
